src/audio/google/google_ctc_audio_processing.c

```diff
@@ -0,0 +1,574 @@
+			    struct sof_ipc_ctrl_data *cdata)
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+	int ret;
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		ret = comp_data_blob_set_cmd(cd->tuning_handler, cdata);
+		if (ret)
+			return ret;
+		if (comp_is_new_data_blob_available(cd->tuning_handler)) {
+			comp_get_data_blob(cd->tuning_handler, NULL, NULL);
+			cd->reconfigure = true;
+		}
+		return 0;
+	default:
+		comp_err(mod->dev,
+			 "google_ctc_audio_processing_ctrl_set_data(): Only binary controls supported %d",
+			 cdata->cmd);
+		return -EINVAL;
+	}
+}
+
+static int ctc_cmd_get_data(struct processing_module *mod,
+			    struct sof_ipc_ctrl_data *cdata,
+			    size_t max_data_size)
+{
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+
+	comp_info(mod->dev, "google_ctc_audio_processing_ctrl_get_data(): %u", cdata->cmd);
+
+	switch (cdata->cmd) {
+	case SOF_CTRL_CMD_BINARY:
+		return comp_data_blob_get_cmd(cd->tuning_handler, cdata, max_data_size);
+	default:
+		comp_err(mod->dev,
+			 "google_ctc_audio_processing_ctrl_get_data(): Only binary controls supported %d",
+			 cdata->cmd);
+		return -EINVAL;
+	}
+}
+#endif
+
+static int ctc_set_config(struct processing_module *mod, uint32_t param_id,
+			  enum module_cfg_fragment_position pos,
+			  uint32_t data_offset_size,
+			  const uint8_t *fragment,
+			  size_t fragment_size, uint8_t *response,
+			  size_t response_size)
+{
+#if CONFIG_IPC_MAJOR_4
+	struct google_ctc_audio_processing_comp_data *cd = module_get_private_data(mod);
+	int ret;
+
+	switch (param_id) {
+	case SOF_IPC4_SWITCH_CONTROL_PARAM_ID:
+	case SOF_IPC4_ENUM_CONTROL_PARAM_ID:
+		comp_err(mod->dev, "google_ctc_audio_processing_set_data(): Only binary controls supported");
+		return -EINVAL;
+	}
+
+	ret = comp_data_blob_set(cd->tuning_handler, pos, data_offset_size,
+				 fragment, fragment_size);
+	if (ret)
+		return ret;
+
+	if (comp_is_new_data_blob_available(cd->tuning_handler)) {
+		comp_get_data_blob(cd->tuning_handler, NULL, NULL);
+		cd->reconfigure = true;
+	}
+
+	return 0;
+#elif CONFIG_IPC_MAJOR_3
+	struct sof_ipc_ctrl_data *cdata = (struct sof_ipc_ctrl_data *)fragment;
+
+	return ctc_cmd_set_data(mod, cdata);
+#endif
+}
+
+static int ctc_get_config(struct processing_module *mod,
+						  uint32_t param_id, uint32_t *data_offset_size,
+						  uint8_t *fragment, size_t fragment_size)
+{
+#if CONFIG_IPC_MAJOR_4
+	comp_err(mod->dev, "ctc_get_config(): Not supported");
+	return -EINVAL;
+#elif CONFIG_IPC_MAJOR_3
+	struct sof_ipc_ctrl_data *cdata = (struct sof_ipc_ctrl_data *)fragment;
+
+	return ctc_cmd_get_data(mod, cdata, fragment_size);
+#endif
+}
+
+static struct module_interface google_ctc_audio_processing_interface = {
+	.init  = ctc_init,
+	.free = ctc_free,
+	.process_audio_stream = ctc_process,
+	.prepare = ctc_prepare,
+	.set_configuration = ctc_set_config,
+	.get_configuration = ctc_get_config,
```

Comment:
Just to repeat from elsewhere: the pattern would be you have a xxx-ipc3.c and xxx-ipc4.c, each of which implements global "google_ctc_s/get_config()" functions which you then put here.  Then you select which one to build in cmake.

Reply:
Done. Thanks!